zerver/lib/push_notifications.py

```diff
@@ -466,77 +459,76 @@ def send_android_push_notification(
     """
     if not devices:
         return 0
-    if not gcm_client:
+    if not fcm_app:
         logger.debug(
-            "Skipping sending a GCM push notification since "
-            "PUSH_NOTIFICATION_BOUNCER_URL and ANDROID_GCM_API_KEY are both unset"
+            "Skipping sending a FCM push notification since "
+            "PUSH_NOTIFICATION_BOUNCER_URL and ANDROID_FCM_CREDENTIALS_PATH are both unset"
         )
         return 0
 
     if remote:
         logger.info(
-            "GCM: Sending notification for remote user %s:%s to %d devices",
+            "FCM: Sending notification for remote user %s:%s to %d devices",
             remote.uuid,
             user_identity,
             len(devices),
         )
     else:
         logger.info(
-            "GCM: Sending notification for local user %s to %d devices", user_identity, len(devices)
+            "FCM: Sending notification for local user %s to %d devices", user_identity, len(devices)
         )
-    reg_ids = [device.token for device in devices]
+
+    token_list = [device.token for device in devices]
     priority = parse_fcm_options(options, data)
-    try:
-        # See https://firebase.google.com/docs/cloud-messaging/http-server-ref .
-        # Two kwargs `retries` and `session` get eaten by `json_request`;
-        # the rest pass through to the FCM server.
-        #
-        # One initial request plus 2 retries, with 5-second timeouts,
-        # and expected 1 + 2 seconds (the gcm module jitters its
-        # backoff by ±50%, so worst case * 1.5) between them, totals
-        # 18s expected, up to 19.5s worst case.
-        res = gcm_client.json_request(
-            registration_ids=reg_ids,
-            priority=priority,
-            data=data,
-            retries=2,
-            session=FCMSession(),
+
+    # The API requires all values to be strings. Our data dict is going to have
+    # things like an integer realm and user ids etc., so just convert everything
+    # like that.
+    data = {k: str(v) if not isinstance(v, str) else v for k, v in data.items()}
```

Comment:
Can we change `data` to `Dict[str, str]` and make the callers stringify things instead, so we know exactly what’s getting stringified?

How and where do these get un-stringified?

Reply:
It could be done, but not that easy to do it in a meaningful way. We could just run this dict-stringifying line in the caller to get it out of here, but that doesn't make anything better because it's still the same "we dont know what's getting stringified" just in a different place.

The meaningful approach would be to improve typing on all the functions that create the payloads to the `send_android_push_notification`, `send_apple_push_notification` functions. I think they're always just `Dict[str, Union[str, int]` so we should be able to at least improve those function signatures with that. And then have that type here for `data` - and then we'd know that the worst thing we can stringifying here is just an `int`.

Or even better, define proper `TypedDict`s for all these functions, getting an exact specification on what the payloads can be.

So I'd rather do that type cleanup after this API migration rather than as part of it, since it's not that simple.

As for what happens with this data later, this goes from [send_each](https://github.com/firebase/firebase-admin-python/blob/32b900b0b7f9d8f453f821bf684bf7023c799808/firebase_admin/messaging.py#L441) -> `_message_data()` -> `_MessagingService.encode_message` -> `JSON_ENCODER.default(message)` [here](https://github.com/firebase/firebase-admin-python/blob/32b900b0b7f9d8f453f821bf684bf7023c799808/firebase_admin/_messaging_encoder.py#L681) where it demands everything in `data`  to be strings without further commentary.
`python-gcm` used to just do a full `json.dumps` on the entire thing given to `json_request()` around [here](https://github.com/geeknam/python-gcm/blob/e958d9128a01dd828221ea5034c66ebbf1a80ea6/gcm/gcm.py#L268)

Reply:
`json.dumps` is quite different from `str`. `json.dumps('3') == '"3"'` is distinguishable from `json.dumps(3) == '3'`, so `json.dumps` is reversible, while `str('3') == str(3) == '3'` is lossy. If whatever’s on the other end of this is expecting JSON, I don’t see how this could be a correct change.

Reply:
I believe it's just that the `data` structure that is passed to the API is supposed to have keys and values as strings, and that that's why stringifying the `int`s is needed. See https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages#Message
```
  "data": {
    string: string,
    ...
  },
```
So there doesn't seem to be any choice in the matter of what else to do with the integer values we send in the payload (like `realm_id` or `user_id`).

I'm not sure whether the result actually looks any different on the mobile app side when the notification comes in; but it is parsed correctly, as evidenced by the fact that these notification gets correctly displayed by the app. Maybe @gnprice may have more insight on the mobile app end of this

Reply:
Here's an example, in our client-side tests, of what the data we get on the client side looks like:
https://github.com/zulip/zulip-flutter/blob/53a45376dfda75374c301428e807a8067685a1c8/test/api/notifications_test.dart#L27-L52
(Or here's the [zulip-mobile version](https://github.com/zulip/zulip-mobile/blob/e352f563ecf2fa9b09b688d5a65b6bc89b0358bc/android/app/src/test/java/com/zulipmobile/notifications/FcmMessageTest.kt), with basically the same information; I largely transcribed that test file from Kotlin to Dart for zulip-flutter.)

So in particular it looks like:
```dart
      "sender_full_name": "A Sender",

      "time": "1546300800",
      "zulip_message_id": "12345",
```

Some of the binding libraries — including FCM / Firebase upstream's own binding library for Dart, [`firebase_messaging`](https://pub.dev/packages/firebase_messaging), which we use in zulip-flutter — describe the data as if it came from JSON. But it's really a string-string map.

Reply:
Given that python-gcm has that `json.dumps` in it, if we are passing it dicts where some of the values are ints and that's actually working, then probably what's going on is that the FCM service has chosen to go Postel's Law on this, so that when it gets `{"a": "3", "b": 4}` it treats it as if it got `{"a": "3", "b": "4"}`. Because it sounds like the former is what we've been sending it from the server, but the latter is definitely what we're seeing on the client.

Reply:
Anyway, in short, switching to `str` here sounds like the right plan to me, because it matches the actual documented API and avoids triggering their use of Postel's Law. Thanks @andersk for raising the question — definitely glad to have looked into it explicitly.

I agree it'd be a great improvement to make the types in the server code more explicit. I don't have a strong opinion on sequencing that or on exactly how the code should then be arranged.

Reply:
Yeah, that's my reading of this too. Just to be precise on this:
> then probably what's going on is that the FCM service has chosen to go Postel's Law on this (...) switching to str here (,,,) avoids triggering their use of Postel's Law.

That seems true for the legacy API. The new API just throws an exception inside `send_each` if we don't `str()` the integers. So they actually force conforming to the documented format.

Reply:
> So they actually force conforming to the documented format.

Cool, good to know.

That's validation that's happening inside the binding library, before sending anything over the network. So it leaves open the question of what the service itself would do if the JSON it gets looks like `…,"time":1546300800,…` — whether it would reject that, or treat it as if it were `…,"time":"1546300800",…`. Either way, though, yeah, since we're using this library, it means we're required to pass strings.